src/target.rs

```diff
@@ -0,0 +1,164 @@
+use arrayvec::ArrayVec;
+use std::io::{BufRead, Cursor};
+use std::process::Output;
+use tokio::process::Command;
+
+/// Compiled target triple, used as default for binary fetching
+pub const TARGET: &str = env!("TARGET");
+
+/// Detect the targets supported at runtime,
+/// which might be different from `TARGET` which is detected
+/// at compile-time.
+///
+/// Return targets supported in the order of preference.
+/// If target_os is linux and it support gnu, then it is preferred
+/// to musl.
+///
+/// If target_os is mac and it is aarch64, then aarch64 is preferred
+/// to x86_64.
+///
+/// Check [this issue](https://github.com/ryankurte/cargo-binstall/issues/155)
+/// for more information.
+pub async fn detect_targets() -> ArrayVec<Box<str>, 2> {
+    if let Some(target) = get_targets_from_rustc().await {
+        let mut v = from_array([target]);
+
+        #[cfg(target_os = "linux")]
+        if v[0].contains("gnu") {
+            v.push(v[0].replace("gnu", "musl").into_boxed_str());
+        }
+
+        #[cfg(target_os = "macos")]
+        if &*v[0] == macos::AARCH64 {
+            v.push(macos::X86.into());
+        }
+
+        v
+    } else {
+        #[cfg(target_os = "linux")]
+        {
+            linux::detect_targets_linux().await
+        }
+        #[cfg(target_os = "macos")]
+        {
+            macos::detect_targets_macos()
+        }
+        #[cfg(not(any(target_os = "linux", target_os = "macos")))]
+        {
+            from_array([TARGET.into()])
+        }
+    }
+}
+
+// Figure out what the host target is, from rustc or from this program's own build target
+async fn get_targets_from_rustc() -> Option<Box<str>> {
```

Comment:
Nitpick: should be singular `target`

Reply:
I think `Option<Box<str>>` is better here since `rustc` might be absense and `get_targets_from_rustc` will fail.
IMO `Option` expresses better than `Targets`, which are a `vec` and it is expected that `Targets` contains at least one target.

Reply:
Sure, yeah, but all I'm saying is that if this is zero-or-one it should be named `get_target_from_rustc -> Option` instead of `get_targets_from_rustc -> Vec`.

Reply:
Got it.

I will update doc of it and rename it to `get_target_from_rustc`.

Reply:
Done!